### PR TITLE
Prometheus application metrics

### PR DESCRIPTION
**Pull request by cableman**
For the reviewer https://prometheus.io/docs/practices/naming/ ... not sure that i got it right.

### CODE REVIEW

File: src/Service/MetricsService.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+     * @param $name
+     *   The name of the metrics
+     * @param $help
+     *   Helper text for the matrices
+     * @param int $value
+     *   The value to increment with
+     * @param $labels
+     *   Labels to filter by in prometheus. Default empty array.
+     */
+    public function counter($name, $help, $value = 1, $labels = [])
```
Comment (tuj) [inline src/Service/MetricsService.php]: Missing parameter types in file

File: src/Controller/MetricsController.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ * Class MetricsController.
+ */
+class MetricsController extends AbstractController
+{
+    /**
+     * @Route("/metrics", name="metrics")
+     *
+     * Render metrics collected by the application.
+     *
+     * @param metricsService $metricsService
```
Comment (tuj) [inline src/Controller/MetricsController.php]: Type should be MetricsService

File: src/Api/DataProvider/AbstractElasticSearchDataProvider.php [hunk 3]
```diff
@@ -44,16 +46,20 @@ abstract class AbstractElasticSearchDataProvider
...
      *   Logger for statistics
+     * @param metricsService $metricsService
+     *   Log metric information
      * @param EventDispatcherInterface $dispatcher
      *   Symfony Event Dispatcher
      * @param IdentifierFactory $factory
      *   Factory to create Identifier Data Transfer Objects (DTOs)
      */
-    public function __construct(Type $index, RequestStack $requestStack, LoggerInterface $statsLogger, EventDispatcherInterface $dispatcher, IdentifierFactory $factory)
+    public function __construct(Type $index, RequestStack $requestStack, LoggerInterface $statsLogger,
```
Comment (tuj) [inline src/Api/DataProvider/AbstractElasticSearchDataProvider.php]: Should be new line for each argument

```
public function __construct(
    Type $index,
    ...
)
{
```

File: src/Api/DataProvider/AbstractElasticSearchDataProvider.php [hunk 4]
```diff
@@ -44,16 +46,20 @@ abstract class AbstractElasticSearchDataProvider
      *   Symfony request stack
      * @param LoggerInterface $statsLogger
      *   Logger for statistics
+     * @param metricsService $metricsService
```
Comment (tuj) [inline src/Api/DataProvider/AbstractElasticSearchDataProvider.php]: Type should be MetricsService

File: src/Service/MoreInfoService/AbstractMoreInfoService.php [hunk 5]
```diff
@@ -84,6 +86,8 @@ abstract class AbstractMoreInfoService extends SoapClient
      *   Elastica index
      * @param LoggerInterface $statsLogger
      *   Statistics logger
+     * @param metricsService $metricsService
```
Comment (tuj) [inline src/Service/MoreInfoService/AbstractMoreInfoService.php]: Type should be MetricsService

File: src/Service/MoreInfoService/AbstractMoreInfoService.php [hunk 6]
```diff
@@ -225,7 +230,8 @@ public function assignSoapHeader(string $headerName, $rawHeaderData = null, stri
      */
     public function moreInfo($body): MoreInfoResponse
     {
-        $start = microtime(true);
+        $totalTime = microtime(true);
+        $labels = ['type' => 'soapRequest'];
```
Comment (tuj) [inline src/Service/MoreInfoService/AbstractMoreInfoService.php]: shouldn't this be moreInfo?

File: src/Service/MoreInfoService/AbstractMoreInfoService.php [hunk 7]
```diff
@@ -235,71 +241,32 @@ public function moreInfo($body): MoreInfoResponse
...
             'service' => 'MoreInfoService',
             'clientID' => $body->authentication->authenticationGroup,
             'remoteIP' => $this->requestStack->getCurrentRequest()->getClientIp(),
             'searchParameters' => $searchParameters,
             'fileNames' => $this->getImageUrls($results),
-            'ElasticQueryTime' => $this->elasticQueryTime,
+            'elasticQueryTime' => $this->elasticQueryTime,
         ]);
-        $this->statsTime = microtime(true) - $statsStart;
+        $this->metricsService->histogram('stats_logging_time', 'Time used logging stats', microtime(true) - $time, $labels);
```
Comment (tuj) [inline src/Service/MoreInfoService/AbstractMoreInfoService.php]: "Prometheus does not have any units hard coded. For better compatibility, base units should be used."
Maybe this should be 'stats_logging_duration_seconds' 

Spelling
'Time used to log stats'

File: src/Service/MoreInfoService/AbstractMoreInfoService.php [hunk 8]
```diff
@@ -235,71 +241,32 @@ public function moreInfo($body): MoreInfoResponse
...
         $response = $this->buildSoapResponse($searchParameters, $results);
 
-        $nohitsStart = microtime(true);
+        $time = microtime(true);
         $this->registerSearchNoHits($response->identifierInformation);
-        $this->nohitsTime = microtime(true) - $nohitsStart;
+        $this->metricsService->histogram('no_hit_event_time', 'Time used to registry no-hit event', microtime(true) - $time, $labels);
 
-        $this->totalTime = microtime(true) - $start;
+        $this->metricsService->histogram('request_time_total', 'Total time used to handel soap request', microtime(true) - $totalTime, $labels);
```
Comment (tuj) [inline src/Service/MoreInfoService/AbstractMoreInfoService.php]: handle

File: src/Service/MoreInfoService/AbstractMoreInfoService.php [hunk 9]
```diff
@@ -235,71 +241,32 @@ public function moreInfo($body): MoreInfoResponse
...
-        $this->statsTime = microtime(true) - $statsStart;
+        $this->metricsService->histogram('stats_logging_time', 'Time used logging stats', microtime(true) - $time, $labels);
 
         $response = $this->buildSoapResponse($searchParameters, $results);
 
-        $nohitsStart = microtime(true);
+        $time = microtime(true);
         $this->registerSearchNoHits($response->identifierInformation);
-        $this->nohitsTime = microtime(true) - $nohitsStart;
+        $this->metricsService->histogram('no_hit_event_time', 'Time used to registry no-hit event', microtime(true) - $time, $labels);
```
Comment (tuj) [inline src/Service/MoreInfoService/AbstractMoreInfoService.php]: to register

File: src/Service/MoreInfoService/AbstractMoreInfoService.php [hunk 10]
```diff
@@ -235,71 +241,32 @@ public function moreInfo($body): MoreInfoResponse
         $boolQuery = $this->buildElasticQuery($searchParameters);
         $search = $this->index->search($boolQuery);
 
-        $this->elasticQueryTime = $search->getResponse()->getQueryTime();
+        $this->metricsService->histogram('elastica_query_time', 'Time used to run elasticsearch query', $search->getResponse()->getQueryTime(), $labels);
```
Comment (tuj) [inline src/Service/MoreInfoService/AbstractMoreInfoService.php]: 'elastica_query_duration_seconds'

File: src/Service/MoreInfoService/AbstractMoreInfoService.php (hunk 9 above)
Comment (tuj) [reply src/Service/MoreInfoService/AbstractMoreInfoService.php]: 'no_hit_event_duration_seconds'

File: src/Service/MoreInfoService/AbstractMoreInfoService.php (hunk 8 above)
Comment (tuj) [reply src/Service/MoreInfoService/AbstractMoreInfoService.php]: 'request_duration_total_seconds'

File: src/Service/MoreInfoService/AbstractMoreInfoService.php [hunk 11]
```diff
@@ -225,7 +230,8 @@ public function assignSoapHeader(string $headerName, $rawHeaderData = null, stri
      */
     public function moreInfo($body): MoreInfoResponse
     {
-        $start = microtime(true);
+        $totalTime = microtime(true);
```
[the remark this answers is not in the document]
Comment (cableman) [reply src/Service/MoreInfoService/AbstractMoreInfoService.php]: Fixed

File: src/Service/MoreInfoService/AbstractMoreInfoService.php [hunk 12]
```diff
@@ -95,12 +99,13 @@ abstract class AbstractMoreInfoService extends SoapClient
      *
      * @throws SoapFault
      */
-    public function __construct(Type $index, LoggerInterface $statsLogger, RequestStack $requestStack,
-                                EventDispatcherInterface $dispatcher, CoverStoreTransformationInterface $transformer,
-                                array $options = [])
+    public function __construct(Type $index, LoggerInterface $statsLogger, MetricsService $metricsService,
+                                RequestStack $requestStack, EventDispatcherInterface $dispatcher,
+                                CoverStoreTransformationInterface $transformer, array $options = [])
```
[the remark this answers is not in the document]
Comment (cableman) [reply src/Service/MoreInfoService/AbstractMoreInfoService.php]: yes and no, phpmd says that it should be one longe line...

File: src/Service/MetricsService.php [hunk 13]
```diff
@@ -0,0 +1,121 @@
+<?php
+/**
+ * @file
+ * Service to help collecting metrics and render them.
```
[the remark this answers is not in the document]
Comment (cableman) [reply src/Service/MetricsService.php]: Fixed

File: src/Service/MetricsService.php (hunk 1 above)
Comment (cableman) [reply src/Service/MetricsService.php]: Fixed

File: src/Controller/MetricsController.php (hunk 2 above)
Comment (cableman) [reply src/Controller/MetricsController.php]: Fixed

File: src/Api/DataProvider/AbstractElasticSearchDataProvider.php (hunk 3 above)
Comment (cableman) [reply src/Api/DataProvider/AbstractElasticSearchDataProvider.php]: Not according to the phpmd tester

File: src/Api/DataProvider/AbstractElasticSearchDataProvider.php (hunk 4 above)
Comment (cableman) [reply src/Api/DataProvider/AbstractElasticSearchDataProvider.php]: Fixed

File: src/Service/MoreInfoService/AbstractMoreInfoService.php (hunk 5 above)
Comment (cableman) [reply src/Service/MoreInfoService/AbstractMoreInfoService.php]: Fixed

File: src/Service/MoreInfoService/AbstractMoreInfoService.php (hunk 6 above)
Comment (cableman) [reply src/Service/MoreInfoService/AbstractMoreInfoService.php]: The type should not (I think).

File: src/Service/MoreInfoService/AbstractMoreInfoService.php (hunk 7 above)
Comment (cableman) [reply src/Service/MoreInfoService/AbstractMoreInfoService.php]: Fixed

File: src/Service/MoreInfoService/AbstractMoreInfoService.php (hunk 8 above)
Comment (cableman) [reply src/Service/MoreInfoService/AbstractMoreInfoService.php]: Fixed

File: src/Service/MoreInfoService/AbstractMoreInfoService.php (hunk 9 above)
Comment (cableman) [reply src/Service/MoreInfoService/AbstractMoreInfoService.php]: Fixed

File: src/Service/MoreInfoService/AbstractMoreInfoService.php (hunk 10 above)
Comment (cableman) [reply src/Service/MoreInfoService/AbstractMoreInfoService.php]: Fixed

File: src/Api/DataProvider/AbstractElasticSearchDataProvider.php (hunk 3 above)
Comment (tuj) [reply src/Api/DataProvider/AbstractElasticSearchDataProvider.php]: Then make it a single line

**Pull request merged**